### PR TITLE
fix some gc rooting

### PR DESCRIPTION
**Pull request by carnaval**
found those while poking around. One is a staged function issue, the other is related to precompilation so probably not well tested.

### CODE REVIEW

**Comment by IainNZ** [comment]
Curious from someone who doesn't know much about the GC: does a missing "GC root" just lead to memory leaks, or can nastier things happen if something isn't rooted?

**Comment by elextr** [comment]
A missing root can mean memory can be unexpectedly freed, resulting in dangling pointers and probably comes in your "nastier things" category.

File: src/gf.c [hunk 1]
```diff
@@ -1520,24 +1520,31 @@ jl_function_t *jl_get_specialization(jl_function_t *f, jl_tupletype_t *types)
...
     if (sf == jl_bottom_func) {
-        return NULL;
+        goto not_found;
     }
     if (sf->linfo == NULL || sf->linfo->ast == NULL) {
-        return NULL;
+        goto not_found;
     }
-    if (sf->linfo->inInference) return NULL;
+    if (sf->linfo->inInference) goto not_found:
```
Comment (tkelman) [inline src/gf.c]: this should be a semicolon

**Comment by timholy** [comment]
@IainNZ, a GC root protects a given variable from being garbage collected. If you don't root something, then
```c
x = do_something();      // this allocates a new variable
y = do_something_else(); // also allocates memory
z = combine_x_and_y(x,y);
```
can cause `x` to go "poof" before that call to `combine`. So the pattern is to protect `x` and `y` from collection during intermediate computations; once the final result is ready, then `JL_GC_POP()` removes that protection.

**Comment by IainNZ** [comment]
Oh wow thats bad, for some reason I thought it was inverse, as you can tell from my wording (i.e. that it'd never be collected). So basically wherever these are missed its basically luck that the GC didn't get called?

**Comment by elextr** [comment]
> So basically wherever these are missed its basically luck that the GC didn't get called?

Yes.

**Comment by IainNZ** [comment]
Wouldn't it be worth running the GC all the time then to flush out the bugs? Not on commits going forward necessarily, but right now to retroactively fix bugs? (sorry for all the questions!)

**Comment by tkelman** [comment]
@IainNZ see some discussion on exactly that idea at https://github.com/JuliaLang/julia/pull/11151#issuecomment-99643392

**Comment by elextr** [comment]
Running the GC more often may help to find such problems, but it is no guarantee, using the example @timholy gave above, even if the unrooted memory is freed before the combination step, if it hasn't been re-allocated and its value still exists, then combination may give the right answer even though the memory has been freed.  But a different scenario may see it re-allocated and its value changed, so combination will give the wrong answer.

Welcome to debugging memory management :)

**Comment by yuyichao** [comment]
@elextr In this case, would clearing out the memory being freed or fill in some invalid pointer values help exposing/tracking down those problem? It will probably have a even deeper impact on the performance but I guess shouldn't be much worse than a always-full-gc approch...

**Comment by JeffBezanson** [comment]
We do that with MEMDEBUG enabled. And I have definitely resorted to increasing GC frequency to try to find bugs. In extreme cases, you can do a collection after every allocation.

**Pull request merged**